# Data Analytics — Exercise 7.3: Pandas DataFrames & Merging
## Ndifon Caleb Ndakina | Date: 5/20/2026

This notebook reads customer order and shipment CSV files into Pandas DataFrames.
The notebook cleans, merges, and analyzes the data to create a complete order fulfilment report for GlobalTrade Ltd.

## 1. Import Pandas and Check Files
This section imports pandas and verifies that the CSV files exist in the current folder

In [1]:
import pandas as pd
import os

print(os.listdir('.'))

['.ipynb_checkpoints', 'A_Exercise_7_2.ipynb', 'DA_Exercise_7_1.ipynb', 'DA_Exercise_7_3.ipynb', 'DA_Guide_7_5.ipynb', 'demo2.ipynb', 'Ex1A_NumPy_Arrays.ipynb', 'Ex1B_NumPy_Random.ipynb', 'Ex1C_Operations_on_Arrays.ipynb', 'import numpy as np.py', 'orders.csv', 'sakila_example.csv', 'shipments.csv', 'Untitled.ipynb', 'Untitled1.ipynb']


## 2. Load CSV Files
This section loads orders.csv and shipments.csv into Pandas DataFrames.

In [3]:
df_orders = pd.read_csv('orders.csv')

df_shipments = pd.read_csv('shipments.csv')

## 3. Preview the Orders DataFrame
Display the first 5 rows, shape, data types, and missing values

In [4]:
print(df_orders.head())

print("\nSHAPE")
print(df_orders.shape)

print("\nDTYPES")
print(df_orders.dtypes)

print("\nMISSING VALUES")
print(df_orders.isnull().sum())

   order_id customer   product  quantity  unit_price
0       101     Chen    Laptop         2      899.99
1       102    Patel    Tablet         5      499.50
2       103   Müller   Monitor         1      349.00
3       104  Okonkwo  Keyboard        10       79.99
4       105   Santos   Headset         3      149.99

SHAPE
(10, 5)

DTYPES
order_id        int64
customer          str
product           str
quantity        int64
unit_price    float64
dtype: object

MISSING VALUES
order_id      0
customer      0
product       0
quantity      0
unit_price    0
dtype: int64


## 4. Preview the Shipments DataFrame
Display the first 5 rows, shape, data types, and missing values.

In [5]:
print(df_shipments.head())

print("\nSHAPE")
print(df_shipments.shape)

print("\nDTYPES")
print(df_shipments.dtypes)

print("\nMISSING VALUES")
print(df_shipments.isnull().sum())

   order_id  ship_date carrier      status
0       101  1/10/2024     DHL   Delivered
1       102  1/12/2024   FedEx   Delivered
2       103  1/13/2024     DHL  In Transit
3       104  1/14/2024     UPS   Delivered
4       105  1/15/2024   FedEx     Delayed

SHAPE
(8, 4)

DTYPES
order_id     int64
ship_date      str
carrier        str
status         str
dtype: object

MISSING VALUES
order_id     0
ship_date    0
carrier      0
status       0
dtype: int64


## 5. Create order_value Column
This section calculates the total value of each order.

In [6]:
df_orders['order_value'] = (
    df_orders['quantity'] *
    df_orders['unit_price']
)

print(df_orders)

   order_id  customer          product  quantity  unit_price  order_value
0       101      Chen           Laptop         2      899.99      1799.98
1       102     Patel           Tablet         5      499.50      2497.50
2       103    Müller          Monitor         1      349.00       349.00
3       104   Okonkwo         Keyboard        10       79.99       799.90
4       105    Santos          Headset         3      149.99       449.97
5       106   Larsson           Webcam         4       89.99       359.96
6       107       Kim            Mouse         8       29.99       239.92
7       108   O'Brien  Docking Station         1      199.00       199.00
8       109     Rossi              SSD         6      119.99       719.94
9       110  Nakamura          USB Hub         2       39.99        79.98


## 6. Numeric Summary Statistics
Use describe() to display summary statistics.

In [7]:
print(df_orders.describe())

        order_id   quantity  unit_price  order_value
count   10.00000  10.000000   10.000000    10.000000
mean   105.50000   4.200000  245.743000   749.515000
std      3.02765   3.047768  273.023774   787.066644
min    101.00000   1.000000   29.990000    79.980000
25%    103.25000   2.000000   82.490000   267.190000
50%    105.50000   3.500000  134.990000   404.965000
75%    107.75000   5.750000  311.500000   779.910000
max    110.00000  10.000000  899.990000  2497.500000


## 7. Convert ship_date to Datetime
This converts the ship_date column from object/string into datetime format.

In [8]:
df_shipments['ship_date'] = pd.to_datetime(
    df_shipments['ship_date']
)

print(df_shipments.dtypes)

order_id              int64
ship_date    datetime64[us]
carrier                 str
status                  str
dtype: object


## 8. Total Order Value
Calculate the total order value for all customer orders.

In [9]:
total = df_orders['order_value'].sum()

print(f"Total order value: £{total:,.2f}")

Total order value: £7,495.15


## 9. INNER MERGE
Only matching order_id values from both DataFrames are kept.

In [10]:
df_inner = pd.merge(
    df_orders,
    df_shipments,
    on='order_id',
    how='inner'
)

print(df_inner)

print("\nSHAPE")
print(df_inner.shape)

   order_id customer   product  quantity  unit_price  order_value  ship_date  \
0       101     Chen    Laptop         2      899.99      1799.98 2024-01-10   
1       102    Patel    Tablet         5      499.50      2497.50 2024-01-12   
2       103   Müller   Monitor         1      349.00       349.00 2024-01-13   
3       104  Okonkwo  Keyboard        10       79.99       799.90 2024-01-14   
4       105   Santos   Headset         3      149.99       449.97 2024-01-15   
5       106  Larsson    Webcam         4       89.99       359.96 2024-01-16   
6       107      Kim     Mouse         8       29.99       239.92 2024-01-18   

  carrier      status  
0     DHL   Delivered  
1   FedEx   Delivered  
2     DHL  In Transit  
3     UPS   Delivered  
4   FedEx     Delayed  
5     UPS   Delivered  
6     DHL  In Transit  

SHAPE
(7, 9)


## 10. LEFT MERGE
Keep all orders even if shipment information is missing.

In [11]:
df_left = pd.merge(
    df_orders,
    df_shipments,
    on='order_id',
    how='left'
)

print(df_left)

   order_id  customer          product  quantity  unit_price  order_value  \
0       101      Chen           Laptop         2      899.99      1799.98   
1       102     Patel           Tablet         5      499.50      2497.50   
2       103    Müller          Monitor         1      349.00       349.00   
3       104   Okonkwo         Keyboard        10       79.99       799.90   
4       105    Santos          Headset         3      149.99       449.97   
5       106   Larsson           Webcam         4       89.99       359.96   
6       107       Kim            Mouse         8       29.99       239.92   
7       108   O'Brien  Docking Station         1      199.00       199.00   
8       109     Rossi              SSD         6      119.99       719.94   
9       110  Nakamura          USB Hub         2       39.99        79.98   

   ship_date carrier      status  
0 2024-01-10     DHL   Delivered  
1 2024-01-12   FedEx   Delivered  
2 2024-01-13     DHL  In Transit  
3 2024-01-14

## 11. Find Unshipped Orders
Rows with missing carrier values are unshipped orders.

In [12]:
unshipped = df_left[
    df_left['carrier'].isna()
]

print(unshipped)

   order_id  customer          product  quantity  unit_price  order_value  \
7       108   O'Brien  Docking Station         1      199.00       199.00   
8       109     Rossi              SSD         6      119.99       719.94   
9       110  Nakamura          USB Hub         2       39.99        79.98   

  ship_date carrier status  
7       NaT     NaN    NaN  
8       NaT     NaN    NaN  
9       NaT     NaN    NaN  


## 12. RIGHT MERGE
Keep all shipment rows even if order information is missing.

In [14]:
df_right = pd.merge(
    df_orders,
    df_shipments,
    on='order_id',
    how='right'
)

print(df_right)

   order_id customer   product  quantity  unit_price  order_value  ship_date  \
0       101     Chen    Laptop       2.0      899.99      1799.98 2024-01-10   
1       102    Patel    Tablet       5.0      499.50      2497.50 2024-01-12   
2       103   Müller   Monitor       1.0      349.00       349.00 2024-01-13   
3       104  Okonkwo  Keyboard      10.0       79.99       799.90 2024-01-14   
4       105   Santos   Headset       3.0      149.99       449.97 2024-01-15   
5       106  Larsson    Webcam       4.0       89.99       359.96 2024-01-16   
6       107      Kim     Mouse       8.0       29.99       239.92 2024-01-18   
7       111      NaN       NaN       NaN         NaN          NaN 2024-01-19   

  carrier      status  
0     DHL   Delivered  
1   FedEx   Delivered  
2     DHL  In Transit  
3     UPS   Delivered  
4   FedEx     Delayed  
5     UPS   Delivered  
6     DHL  In Transit  
7   FedEx   Delivered  


## 13. Find Orphan Shipment
Find shipment rows without matching order records.

In [15]:
orphan = df_right[
    df_right['customer'].isna()
]

print(orphan)

   order_id customer product  quantity  unit_price  order_value  ship_date  \
7       111      NaN     NaN       NaN         NaN          NaN 2024-01-19   

  carrier     status  
7   FedEx  Delivered  


## 14. OUTER MERGE
Keep all rows from both DataFrames.

In [16]:
df_outer = pd.merge(
    df_orders,
    df_shipments,
    on='order_id',
    how='outer'
)

print(df_outer)

print("\nSHAPE")
print(df_outer.shape)

print("\nTOTAL NaN VALUES")
print(df_outer.isnull().sum().sum())

    order_id  customer          product  quantity  unit_price  order_value  \
0        101      Chen           Laptop       2.0      899.99      1799.98   
1        102     Patel           Tablet       5.0      499.50      2497.50   
2        103    Müller          Monitor       1.0      349.00       349.00   
3        104   Okonkwo         Keyboard      10.0       79.99       799.90   
4        105    Santos          Headset       3.0      149.99       449.97   
5        106   Larsson           Webcam       4.0       89.99       359.96   
6        107       Kim            Mouse       8.0       29.99       239.92   
7        108   O'Brien  Docking Station       1.0      199.00       199.00   
8        109     Rossi              SSD       6.0      119.99       719.94   
9        110  Nakamura          USB Hub       2.0       39.99        79.98   
10       111       NaN              NaN       NaN         NaN          NaN   

    ship_date carrier      status  
0  2024-01-10     DHL   Del

## 15. Add Shipped Flag
Create a True/False column showing whether the order shipped.

In [17]:
df_left['shipped'] = (
    df_left['carrier'].notna()
)

print(df_left[
    ['order_id',
     'customer',
     'order_value',
     'shipped']
])

   order_id  customer  order_value  shipped
0       101      Chen      1799.98     True
1       102     Patel      2497.50     True
2       103    Müller       349.00     True
3       104   Okonkwo       799.90     True
4       105    Santos       449.97     True
5       106   Larsson       359.96     True
6       107       Kim       239.92     True
7       108   O'Brien       199.00    False
8       109     Rossi       719.94    False
9       110  Nakamura        79.98    False


## 16. Count Shipped vs Unshipped Orders

In [18]:
print(
    df_left['shipped'].value_counts()
)

shipped
True     7
False    3
Name: count, dtype: int64


## 17. Fulfilment Rate
Calculate the percentage of shipped orders.

In [19]:
shipped_count = df_left['shipped'].sum()

total_orders = len(df_left)

fulfilment_rate = (
    shipped_count /
    total_orders * 100
)

print(
    f"Fulfilment rate: {fulfilment_rate:.2f}%"
)

Fulfilment rate: 70.00%


## 18. GroupBy Summary by Status
Analyze order values by shipment status

In [20]:
status_summary = (
    df_inner.groupby('status')
    ['order_value']
    .agg(['sum', 'mean', 'count'])
)

print(status_summary)

                sum      mean  count
status                              
Delayed      449.97   449.970      1
Delivered   5457.34  1364.335      4
In Transit   588.92   294.460      2


## 19. Carrier Totals
Display total order values by carrier

In [21]:
carrier_totals = (
    df_inner.groupby('carrier')
    ['order_value']
    .sum()
    .sort_values(ascending=False)
)

print(carrier_totals)

print(
    f"\nTop carrier: {carrier_totals.index[0]}"
)

carrier
FedEx    2947.47
DHL      2388.90
UPS      1159.86
Name: order_value, dtype: float64

Top carrier: FedEx


## 20. Units Dispatched by Status
Display total quantity shipped for each status.

In [22]:
units_per_status = (
    df_inner.groupby('status')
    ['quantity']
    .sum()
)

print(units_per_status)

status
Delayed        3
Delivered     21
In Transit     9
Name: quantity, dtype: int64


##  See Summary Below

- The fulfilment rate is 70%, meaning 7 out of 10 orders were shipped.
- FedEx had the highest total shipment value.
- Order ID 111 is an orphan shipment because it does not have a matching order record.